c/src/atom.rs

```diff
@@ -644,49 +649,141 @@ pub unsafe extern "C" fn atom_eq(atoma: *const atom_ref_t, atomb: *const atom_re
     (&*atoma).borrow() == (&*atomb).borrow()
 }
 
-// TODO for Alpha: Unify vec_atom_t with atom_array_t
-pub struct vec_atom_t(pub(crate) Vec<Atom>);
+/// Represents a list (aka Vec) of Atoms.  It is unsafe to directly access the fields of this struct, so
+/// accessors should be used instead
+#[repr(C)]
+pub struct atom_vec_t {
+    ptr: *mut RustOpaqueAtom,
+    len: usize,
+    capacity: usize,
+    owned: bool,
+}
+
+impl atom_vec_t {
+    fn new() -> Self {
+        Vec::<Atom>::new().into()
+    }
+    fn as_slice(&self) -> &[Atom] {
+        core::borrow::Borrow::borrow(self)
+    }
+    /// Converts a borrowed vec into an owned vec
+    fn into_owned(self) -> Self {
+        match self.owned {
+            true => self,
+            false => {
+                self.as_slice().to_vec().into()
+            }
+        }
+    }
+}
+
+impl From<Vec<Atom>> for atom_vec_t {
+    fn from(vec: Vec<Atom>) -> Self {
+        //When Vec::into_raw_parts is stabilized then use it.  https://github.com/rust-lang/rust/issues/65816
+        let mut vec = core::mem::ManuallyDrop::new(vec);
+        Self {
+            ptr: vec.as_mut_ptr().cast(),
+            len: vec.len(),
+            capacity: vec.capacity(),
+            owned: true,
+        }
+    }
+}
+
+impl From<&[Atom]> for atom_vec_t {
+    fn from(slice: &[Atom]) -> Self {
+        Self {
+            ptr: slice.as_ptr().cast_mut().cast(),
+            len: slice.len(),
+            capacity: slice.len(),
+            owned: false,
+        }
+    }
+}
+
+impl From<atom_vec_t> for Vec<Atom> {
+    fn from(vec: atom_vec_t) -> Self {
+        if !vec.owned {
+            panic!("Error! Attempt to take ownership of borrowed atom_vec_t");
+        }
+        let rust_vec = unsafe{ Vec::from_raw_parts(vec.ptr.cast(), vec.len, vec.capacity) };
+        core::mem::forget(vec);
+        rust_vec
+    }
+}
+
+impl Drop for atom_vec_t {
+    fn drop(&mut self) {
+        if self.owned {
+            let vec: Vec<Atom> = unsafe{ Vec::from_raw_parts(self.ptr.cast(), self.len, self.capacity) };
+            drop(vec);
+        }
+    }
+}
+
+impl core::borrow::Borrow<[Atom]> for atom_vec_t {
+    fn borrow(&self) -> &[Atom] {
+        unsafe{ core::slice::from_raw_parts(self.ptr.cast(), self.len) }
+    }
+}
 
 #[no_mangle]
-pub extern "C" fn vec_atom_new() -> *mut vec_atom_t {
-    vec_atom_into_ptr(Vec::new())
+pub extern "C" fn atom_vec_new() -> atom_vec_t {
+    atom_vec_t::new()
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn vec_atom_free(vec: *mut vec_atom_t) {
-    drop(Box::from_raw(vec));
+pub unsafe extern "C" fn atom_vec_from_array(atoms: *mut atom_t, size: usize) -> atom_vec_t {
+    let c_arr = std::slice::from_raw_parts_mut(atoms, size);
+    let atoms: Vec<Atom> = c_arr.into_iter().map(|atom| {
+        core::mem::replace(atom, atom_t::null()).into_inner()
+    }).collect();
+    atoms.into()
```

Comment:
Just curious, is there a difference with the following implementation:
```rust
pub unsafe extern "C" fn atom_vec_from_array(atoms: *mut atom_t, size: usize) -> atom_vec_t {
    let atoms: Vec<atom_t> = Vec::from_raw_parts(atoms, size, size);
    let atoms: Vec<Atom> = atoms.into_iter().map(|atom| atom.into_inner()).collect();
    atoms.into()
}
```

Reply:
The issue is that `Vec::from_raw_parts` has a number of safety considerations that we can't guarantee if we're taking the pointer from C.  Namely which allocator was used, the allocation size (it's not enough that the allocation is big enough - it has to be exact), and the pointer alignment.

So basically, the only time `Vec::from_raw_parts` is safe is if the vec started life as a Rust Vec, and we're just letting C hold onto the pointer.

`slice::from_raw_parts` is a lot more forgiving about the pointer details because it doesn't take responsibility for managing the pointer later on. (freeing / growing / etc.)

Reply:
Ah ok, you are right, missed it in `Vec` documentation. Thanks!